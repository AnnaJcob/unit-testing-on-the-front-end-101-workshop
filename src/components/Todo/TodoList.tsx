// Very simple and rudimentar todo component for a simple unit testing on the front end workshop (101)

import React, { useState, useRef } from 'react'

import styles from './TodoList.module.scss' ;

interface ITodoItemProps {
    itemText: string;
    onRemove: (itemText: string) => void;
}

const TodoItem: React.FC<ITodoItemProps> = props => {
    return (
        <li>
            <span>{props.itemText}</span><a href="#" onClick={() => props.onRemove(props.itemText)}> (Remove)</a>
        </li>
    );
};

const TodoList: React.FC = () => {
    const [newTaskText, setNewTaskText] = useState<string>('');
    const [todoItems, setTodoItems] = useState<string[]>([]);
    const inputRef = useRef<HTMLInputElement>(null);

    const addTodoItem = (event: React.FormEvent<HTMLFormElement>) => {
        event.preventDefault();
        
        if (inputRef.current)
        {
            const itemText = inputRef.current.value;

            if (itemText.trim().length > 0 && todoItems.findIndex(t => t === itemText) < 0)
            {
                setTodoItems([...todoItems, itemText]);
            }

            setNewTaskText('');
        }
    }

    const removeTodoItem = (itemText: string) => {
        const indexToRemove = todoItems.findIndex(t => t === itemText);

        if (indexToRemove >= 0)
            setTodoItems([...todoItems.slice(0, indexToRemove), ...todoItems.slice(indexToRemove + 1)]);
    }

    const newTaskTextUpdated = (event: React.ChangeEvent<HTMLInputElement>) => {
        setNewTaskText(event.target.value);
    }

    return (
        <div className={styles["todo-list"]}>
            <div className={styles["todo-list-header"]}>
                <form onSubmit={addTodoItem}>
                    <input 
                        type="text" 
                        placeholder="Enter your Task" 
                        ref={inputRef} 
                        required
                        onChange={newTaskTextUpdated}
                        value={newTaskText}
                    />
                    <button type="submit">Add Task</button>
                </form>
            </div>
            <ul className={styles["todo-list-items"]}>
                {todoItems.map((item, index) => {
                    return <TodoItem key={index} itemText={item} onRemove={removeTodoItem} />
                })}
            </ul>
        </div>
    );
};

export default TodoList

