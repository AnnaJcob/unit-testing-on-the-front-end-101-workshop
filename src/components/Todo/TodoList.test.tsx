import React from 'react';
import ReactDOM from 'react-dom';
import TodoList from './TodoList';
import { mount } from 'enzyme';

describe("<Todo", () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<TodoList />, div);
    });

    it('should match snapshot', () => {
      const mountedComponent = mount(<TodoList />);
      expect(mountedComponent).toMatchSnapshot();
  });
});
