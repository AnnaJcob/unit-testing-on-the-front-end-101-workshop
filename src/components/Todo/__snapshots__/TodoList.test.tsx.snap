// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todo should match snapshot 1`] = `
"<TodoList>
  <div className=\\"todo-list\\">
    <div className=\\"todo-list-header\\">
      <form onSubmit={[Function: addTodoItem]}>
        <input type=\\"text\\" placeholder=\\"Enter your Task\\" required={true} onChange={[Function: newTaskTextUpdated]} value=\\"\\" />
        <button type=\\"submit\\">
          Add Task
        </button>
      </form>
    </div>
    <ul className=\\"todo-list-items\\" />
  </div>
</TodoList>"
`;
